dkg/dkg.go

```diff
@@ -48,6 +53,104 @@ type Config struct {
 	TestSigning bool
 }
 
+type sigData struct {
+	duty   core.DutyType
+	pubkey core.PubKey
+	psigs  []core.ParSignedData
+}
+
+// exchanger is responsible for exchanging partial signatures between peers on libp2p.
+type exchanger struct {
+	sigChan  chan sigData
+	sigex    *parsigex.ParSigEx
+	sigdb    *parsigdb.MemDB
+	numPeers int
+	numVals  int
+}
+
+func newExchanger(tcpNode host.Host, peerIdx int, peers []peer.ID, vals int) *exchanger {
+	ex := &exchanger{
+		// threshold is len(peers) to wait until we get all the partial sigs from all the peers per DV
+		sigdb:    parsigdb.NewMemDB(len(peers)),
+		sigex:    parsigex.NewParSigEx(tcpNode, peerIdx, peers),
+		sigChan:  make(chan sigData, vals),
+		numPeers: len(peers),
+		numVals:  vals,
+	}
+
+	// Wiring core workflow components
+	ex.sigdb.SubscribeInternal(ex.sigex.Broadcast)
+	ex.sigdb.SubscribeThreshold(ex.pushPsigs)
+	ex.sigex.Subscribe(ex.sigdb.StoreExternal)
+
+	return ex
+}
+
+// exchange exhanges partial signatures of lockhash/deposit-data among dkg participants and returns all the partial
+// signatures of the group according to public key of each DV.
+func (e *exchanger) exchange(ctx context.Context, duty core.DutyType, set core.ParSignedDataSet) ([]core.ParSignedDataSet, error) {
+	// Start the process by storing current peer's ParSignedDataSet
+	err := e.sigdb.StoreInternal(ctx, core.Duty{Type: duty}, set)
+	if err != nil {
+		return nil, err
+	}
+
+	// len(sets) is numPeers + 1 because of shareIdx being 1-indexed
+	sets := make([]core.ParSignedDataSet, e.numPeers+1)
```

Comment:
I would not use slices that are 1-indexed. I would rather:
 - use map: `map[shareIdx]core.ParSignedDataSet`
 - or 0-indexed slice: [peerIdx]core.ParSignedDataSet

Reply:
This would also allow you to count the length of this thing, as opposed to keeping a separate counter (remaining)